### PR TITLE
Upgrade TypeScript, CI config and support Layer type inference.

### PR DESCRIPTION
**Pull request by Losses**
Say, we have a net, and do following operation:

```
const layer = net.layers[0];
if (layer.layer_type === 'conv') {
  layer
}
```

In previous version, layer will always be ILayer, but TypeScript could automatically inference it from its typing.

With this pull request, the type of `layer` will become `ConvLayer`, which is more detailed.

### CODE REVIEW

**Comment by Losses** [comment]
And following commits will improve the typing of "fromJSON" and "toJSON".

**Comment by Losses** [comment]
Folling commits will export useful types for developers.

**Comment by Losses** [comment]
@yoskeoka Hmmmmm... Hello?

File: .travis.yml [hunk 1]
```diff
@@ -1,4 +1,3 @@
 language: node_js
 node_js:
-  - "7"
-  - "6"
+  - "10"
```
Comment (yoskeoka) [inline .travis.yml]: Why wouldn't you use "14" ?
In `package.json` , "@types/node" is specified 14.

**Review by yoskeoka — CHANGES_REQUESTED** [review]
Looks great work.

Please fix some.

- Run CI with node v14 as you specify in `devDependencies`.
- Switch `yarn` to `npm`. Generate lock file with `package-lock.json` filename.
- `test/NeuralNet.spec.ts` throws error with `yarn test` command.

**Comment by Losses** [comment]
@yoskeoka Fixed, please check the code

**Comment by yoskeoka** [comment]
@Losses 
LGTM!!

**Review by yoskeoka — APPROVED** [review]

**Pull request merged**